Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-203.ca-central-1.compute.internal
Sha: 67f1802b4218b3d620cb475ff993cb895078f716
Number of Error(s): 5
/home/centos/smoketest/PR-13515/HPCC-Platform/system/jlib/jptree.cpp:7861:10: error: unused variable \xe2\x80\x98held\xe2\x80\x99 [-Werror=unused-variable]
cc1plus: all warnings being treated as errors
make[2]: *** [system/jlib/CMakeFiles/jlib.dir/build.make:427: system/jlib/CMakeFiles/jlib.dir/jptree.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1266: system/jlib/CMakeFiles/jlib.dir/all] Error 2
make: *** [Makefile:163: all] Error 2

[Build error:](https://gist.github.com/997dee8de57f9b57d899ddd03e21e053)
Build log [Build log](https://gist.github.com/HPCCSmoketest/997dee8de57f9b57d899ddd03e21e053#file-result-20-03-18-11-38-06-log)

--- a/dockerfiles/hpcc/templates/thor.yaml
+++ b/dockerfiles/hpcc/templates/thor.yaml
@@ -32,10 +32,9 @@ spec:
       containers:
       - name: {{ $eclAgentName | quote }}
         args: [
-                {{ include "hpcc.configArg" $eclAgentScope }},
+                {{ include "hpcc.configArg" . }},
                 {{ include "hpcc.daliArg" $ }},
-                {{ printf "\"--name=%s\"" .name }},
-                {{ printf "\"--processConfig=/etc/config/%s.yaml\"" $hthorName }}
+                {{ printf "\"--name=%s\"" .name }}
               ]
 {{- include "hpcc.addSecurityContext" (dict "root" $ "me" .) | indent 8 }}
 {{ include "hpcc.addImageAttrs" (dict "root" $ "me" . "imagename" "eclagent") | indent 8 }}
@@ -66,8 +65,7 @@ spec:
         args: [
                 {{ include "hpcc.configArg" $thorAgentScope }},
                 {{ include "hpcc.daliArg" $ }},
-                {{ printf "\"--name=%s\"" .name }},
-                {{ printf "\"--processConfig=/etc/config/%s.yaml\"" .name }}
+                {{ printf "\"--name=%s\"" .name }}
               ]
 {{- include "hpcc.addSecurityContext" (dict "root" $ "me" .) | indent 8 }}
 {{ include "hpcc.addImageAttrs" (dict "root" $ "me" . "imagename" "eclagent") | indent 8 }}
@@ -87,12 +85,9 @@ data:
     version: 1.0
     thor:
 {{ toYaml $thorScope | indent 6 }}
-    global:
-{{ include "hpcc.generateGlobalConfigMap" $ | indent 6 }}
-
-  {{ $eclAgentScope.name }}.yaml: |
-    version: 1.0
-    eclagent:
+    hthor:
+{{ toYaml $hthorScope | indent 6 }}
+    eclagent: # main agent Q handler
 {{ toYaml $eclAgentScope | indent 6 }}
     global:
 {{ include "hpcc.generateGlobalConfigMap" $ | indent 6 }}
@@ -104,13 +99,7 @@ data:
       type: thor
     global:
 {{ include "hpcc.generateGlobalConfigMap" $ | indent 6 }}
-
-  {{ $hthorScope.name }}.yaml: |
-    version: 1.0
-    hthor:
-{{ toYaml $hthorScope | indent 6 }}
-    global:
-{{ include "hpcc.generateGlobalConfigMap" $ | indent 6 }}
+{{- if $eclAgentScope.containerPerAgent }}
 
   hthor-jobspec.yaml: |
     apiVersion: batch/v1
@@ -136,7 +125,7 @@ data:
                 #limit-cpu
             command: [
                         "hthor",
-                        {{ include "hpcc.configArg" $hthorScope }},
+                        {{ include "hpcc.configArg" . }},
                         {{ include "hpcc.daliArg" $ }},
                         %args
                      ]
@@ -148,6 +137,9 @@ data:
 {{ include "hpcc.addVolumes" $ | indent 10 }}
           restartPolicy: Never
       backoffLimit: 0
+{{- end }}
+{{- if $thorAgentScope.containerPerAgent }}
+
   thormaster-jobspec.yaml: |
     apiVersion: batch/v1
     kind: Job
@@ -164,7 +156,7 @@ data:
 {{- include "hpcc.addSecurityContext" (dict "root" $ "me" .) | indent 12 }}
 {{ include "hpcc.addImageAttrs" (dict "root" $ "me" . "imagename" "thormaster") | indent 12 }}
             args: [
-                    {{ include "hpcc.configArg" $thorScope }},
+                    {{ include "hpcc.configArg" . }},
                     {{ include "hpcc.daliArg" $ }},
                     %args
                   ]
@@ -176,6 +168,8 @@ data:
 {{ include "hpcc.addVolumes" $ | indent 10 }}
           restartPolicy: Never
       backoffLimit: 0
+{{- end }}
+
   thorslave-jobspec.yaml: |
     apiVersion: batch/v1
     kind: Job
@@ -192,7 +186,7 @@ data:
 {{- include "hpcc.addSecurityContext" (dict "root" $ "me" .) | indent 12 }}
 {{ include "hpcc.addImageAttrs" (dict "root" $ "me" . "imagename" "thorslave") | indent 12 }}
             args: [
-                    {{ include "hpcc.configArg" $thorAgentScope }},
+                    {{ include "hpcc.configArg" . }},
                     {{ include "hpcc.daliArg" $ }},
                     %args
                   ]

--- a/ecl/agentexec/agentexec.cpp
+++ b/ecl/agentexec/agentexec.cpp
@@ -248,10 +248,10 @@ public:
             else
             {
                 VStringBuffer exec("%s --workunit=%s --daliServers=%s", processName.get(), wuid.str(), dali.str());
-                if (queryComponentConfig().hasProp("@processConfig"))
+                if (queryComponentConfig().hasProp("@config"))
                 {
                     exec.append(" --config=");
-                    queryComponentConfig().getProp("@processConfig", exec);
+                    queryComponentConfig().getProp("@config", exec);
                 }
                 if (queryComponentConfig().getPropBool("@useThorQueue", true))
                     exec.append(" --queue=").append(queue);

--- a/system/jlib/jptree.cpp
+++ b/system/jlib/jptree.cpp
@@ -24,6 +24,7 @@
 #include "jlzw.hpp"
 #include "jregexp.hpp"
 #include "jstring.hpp"
+#include "jutil.hpp"
 #include "yaml.h"
 
 #ifdef __APPLE__
@@ -31,7 +32,7 @@
 #define environ (*_NSGetEnviron())
 #endif
 
-#include <algorithm>
+#include <initializer_list>
 
 #define MAKE_LSTRING(name,src,length) \
     const char *name = (const char *) alloca((length)+1); \
@@ -7754,16 +7755,8 @@ static const char * extractOption(const char * option, const char * cur)
     return nullptr;
 }
 
-static bool ignoreOption(const char * name)
-{
-    return streq(name, "config") || streq(name, "outputconfig");
-}
-
 static void applyCommandLineOption(IPropertyTree * config, const char * option, const char * value)
 {
-    if (ignoreOption(option))
-        return;
-
     if (islower(*option))
     {
         StringBuffer path;
@@ -7777,20 +7770,25 @@ static void applyCommandLineOption(IPropertyTree * config, const char * option, 
     }
 }
 
-static void applyCommandLineOption(IPropertyTree * config, const char * option)
+static void applyCommandLineOption(IPropertyTree * config, const char * option, std::initializer_list<const char *> ignoreOptions)
 {
     const char * eq = strchr(option, '=');
+    StringBuffer name;
+    const char *val = nullptr;
     if (eq)
     {
-        StringBuffer name;
         name.append(eq - option, option);
-        applyCommandLineOption(config, name, eq + 1);
+        option = name;
+        val = eq + 1;
     }
     else
     {
         //MORE: Support --x- and --x+?
-        applyCommandLineOption(config, option, "1");
+        val = "1";
     }
+    if (stdContains(ignoreOptions, option))
+        return;
+    applyCommandLineOption(config, option, val);
 }
 
 static Owned<IPropertyTree> componentConfiguration;
@@ -7819,13 +7817,13 @@ IPropertyTree & queryGlobalConfig()
     return *globalConfiguration;
 }
 
-jlib_decl IPropertyTree * loadArgsIntoConfiguration(IPropertyTree *config, const char * * argv)
+jlib_decl IPropertyTree * loadArgsIntoConfiguration(IPropertyTree *config, const char * * argv, std::initializer_list<const char *> ignoreOptions)
 {
     for (const char * * pArg = argv; *pArg; pArg++)
     {
         const char * cur = *pArg;
         if (startsWith(cur, "--"))
-            applyCommandLineOption(config, cur + 2);
+            applyCommandLineOption(config, cur + 2, ignoreOptions);
     }
     return config;
 }
@@ -7859,6 +7857,9 @@ jlib_decl IPropertyTree * loadConfiguration(const char * defaultYaml, const char
     Linked<IPropertyTree> config(componentDefault);
     const char * optConfig = nullptr;
     bool outputConfig = false;
+#ifdef _DEBUG
+    bool held = false;
+#endif
     for (const char * * pArg = argv; *pArg; pArg++)
     {
         const char * cur = *pArg;
@@ -7883,11 +7884,14 @@ jlib_decl IPropertyTree * loadConfiguration(const char * defaultYaml, const char
 #ifdef _DEBUG
         else
         {
-            const char * matchHold = extractOption("--hold", cur);
+            const char *matchHold = extractOption("--hold", cur);
             if (matchHold)
             {
                 if (strToBool(matchHold))
+                {
+                    held = true;
                     holdLoop();
+                }
             }
         }
 #endif
@@ -7927,15 +7931,16 @@ jlib_decl IPropertyTree * loadConfiguration(const char * defaultYaml, const char
         applyEnvironmentConfig(*config, envPrefix, *cur);
     }
 
-    loadArgsIntoConfiguration(config, argv);
-
 #ifdef _DEBUG
-    if (config->getPropBool("@hold"))
+    // NB: don't re-hold, if CLI --hold already held.
+    if (!held && config->getPropBool("@hold"))
         holdLoop();
 #endif
 
     if (outputConfig)
     {
+        loadArgsIntoConfiguration(config, argv, { "config", "outputconfig" });
+
         Owned<IPropertyTree> recreated = createPTree();
         recreated->setProp("@version", currentVersion);
         recreated->addPropTree(componentTag, LINK(config));
@@ -7946,6 +7951,8 @@ jlib_decl IPropertyTree * loadConfiguration(const char * defaultYaml, const char
         printf("%s\n", yamlText.str());
         exit(0);
     }
+    else
+        loadArgsIntoConfiguration(config, argv);
 
     if (!globalConfiguration)
         globalConfiguration.setown(createPTree("global"));

--- a/system/jlib/jptree.hpp
+++ b/system/jlib/jptree.hpp
@@ -24,6 +24,8 @@
 #include "jiter.hpp"
 #include "jprop.hpp"
 
+#include <initializer_list>
+
 enum TextMarkupFormat
 {
     MarkupFmt_Unknown=0,
@@ -293,7 +295,7 @@ inline static bool isValidXPathChr(char c)
     return ('\0' != c && (isalnum(c) || strchr(validChrs, c)));
 }
 
-jlib_decl IPropertyTree * loadArgsIntoConfiguration(IPropertyTree *config, const char * * argv);
+jlib_decl IPropertyTree * loadArgsIntoConfiguration(IPropertyTree *config, const char * * argv, std::initializer_list<const char *> ignoreOptions = {});
 jlib_decl IPropertyTree * loadConfiguration(const char * defaultYaml, const char * * argv, const char * componentTag, const char * envPrefix, const char * legacyFilename, IPropertyTree * (mapper)(IPropertyTree *));
 jlib_decl IPropertyTree * queryCostsConfiguration();
 

--- a/system/jlib/jutil.hpp
+++ b/system/jlib/jutil.hpp
@@ -24,6 +24,9 @@
 #include "jarray.hpp"
 #include "jbuff.hpp"
 
+#include <algorithm> 
+#include <iterator>
+
 #if defined (__APPLE__)
 #include <mach/mach_time.h>
 extern mach_timebase_info_data_t timebase_info;   // Calibration for nanosecond timer
@@ -555,6 +558,12 @@ protected:
     unsigned initialDelta;
     bool isQuantile;
 };
+
+template <typename Container, typename Value>
+inline bool stdContains(Container&& container, Value &&v)
+{
+    return container.end() != std::find(container.begin(), container.end(), std::forward<Value>(v));
+}
 
 #endif
 

